[TECH] Verifier les parametre de la route d'ajout de candidat (PIX-8537)

aceol opened the pull request:
## :unicorn: Problème
Il n'y a pas de verification de la route d'ajout de candidat. 

## :robot: Proposition
Valider les parametre via JOI

## :rainbow: Remarques
Ajout de tests sur la suppression de candidat (pas de modification sur le code)

## :100: Pour tester
Ajouter des candidats via la modale d'une session SCO/PRO
(tester avec et sans certification-complementaires)

aceol committed:
:sparkles: api: Check certification-candidate POSTT route parameters

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6517.review.pix.fr)
  * [App (.org)](https://app-pr6517.review.pix.org)
  * [Orga (.fr)](https://orga-pr6517.review.pix.fr)
  * [Orga (.org)](https://orga-pr6517.review.pix.org)
  * [Certif (.fr)](https://certif-pr6517.review.pix.fr)
  * [Certif (.org)](https://certif-pr6517.review.pix.org)
  * [Admin](https://admin-pr6517.review.pix.fr)
  * [API](https://api-pr6517.review.pix.fr/api/)
  * [Pix1D](https://1d-pr6517.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6517/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6517/environment)

aceol committed:
:white_check_mark: api: Add tests to the delete certification candidates route

P-Jeremy commented:
test func ok

![image](https://github.com/1024pix/pix/assets/37305474/554d0cfb-a5d0-4088-8387-ccc6841a046b)

AndreiaPena commented on api/lib/application/sessions/index.js:
```diff
@@ -290,6 +293,37 @@ const register = async function (server) {
             id: identifiersType.sessionId,
           }),
+          payload: Joi.object({
+            data: {
```
**Question :** 
on a pas une validation du certification candidate qui se fait plus tard sur cette route ? (méthode `validate`)

Si oui, est-ce que c'est pertinent d'avoir cette vérif ici sachant que dans notre validation (plus tard) on se charge de retourner les bons code d'erreur ?

Est-ce que ça ne fait pas tout simplement doublon ? (même si je suis d'accord que le joi devrait d'abord jeter ici et pas après)

AndreiaPena replied:
Pour vérifier ça il faut faire un test fonctionnel avec erreur pour voir ce que le front dit.

Steph0 replied:
Je ne trouve pas que cela fait doublon car ça permet de documenter l'API sans devoir aller chercher dans le code

Steph0 replied:
Désormais avec cette PR, si on va sur https://api-pr6517.review.pix.fr/api/swagger.json
sur l'api `/certification-candidates` on peut voir le lien avec le model `Model137` qui comprend désormais la définition de l'objet attendu.

AndreiaPena replied:
si ça a une valeur à documenter alors oui mais dans les faits, cette route fait 2 fois appel à Joi sur les mêmes paramètres non ?

Steph0 replied:
Tu as raison dans les faits, appliqué au front. Mais il faut voir que les API ne sont pas forcement appelées par les front (certains clients contourne le front par exemple, sur pix).

Ajouter du Joi comme dans cette PR permet à un consommateur de savoir quoi envoyer. La difficulté est par contre de s'assurer que l'API n'est pas moins permissive que ce que le métier accepte déjà.

Ici ton test fonc le prouve, l'API ne devrait pas mettre `.required` car cette vérification est une contrainte de la couche métier. On doit `allow(null)`.

Avec `allow(null)` on dit au consommateur (quel qu'il soit) "Il existe une propriété first-name qui doit être renseignée, au pire à null". Sans cette PR, si quelqu'un veut appeler l'API, il n'aurait aucune connaissance d'une propriété `first-name`.

P-Jeremy replied:
Je rejoins @AndreiaPena, j'ai mis la coche mais cette double verif me dérange. Surout si on perd les codes d'erreurs

aceol replied:
La verification coté route se fait maintenant que sur les noms/types de parametres mais pas sur le "required". 
On a plus une (forte) double verification. 
Il y avait aussi un soucis ca le front faisait un trim. On avait donc un soucis lorsque les champs contenait uniquement des espaces. (pas d'erreur coté front, mais erreur de champs required coté back). C'est corrigé (on trim coté front et pas back)

AndreiaPena commented:
C'est le soucis que je craignais, on à ici le joi de la route qui plante mais du coup plus de code d'erreur

<img width="1200" alt="Capture d’écran 2023-07-06 à 17 12 20" src="https://github.com/1024pix/pix/assets/58915422/65e8da5d-f18b-4534-993f-bae4daf2e4e7">

Steph0 commented:
Tech test OK ✔️ 

Comme ça porte sur l'API seulement et que ça ne change pas le payload de sortie j'ai fais le test fonc via un API d'appel API (Postman).

![image](https://github.com/1024pix/pix/assets/170271/e137c059-3431-4433-9122-301cf240fb4d)

aceol committed:
:recycle: certif, api: Check empty values in api

aceol committed:
:bug: certif: Fix PixSelect required error message

PixSelect does not include form html component so if the field is empty and required
there is no native message displayed. Best would be to implement our own error messages but as
a quick fix, we display a hidden input error message
